ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
+        }
+    }
+
+  private[this] val flatTransactionById =
+    LedgerTest("TXFlatTransactionById", "Expose a visible transaction by identifier") { context =>
+      for {
+        ledger <- context.participant()
+        party <- ledger.allocateParty()
+        dummy <- ledger.create(party, Dummy(party))
+        tree <- ledger.exerciseForFlatTransaction(party, dummy.exerciseDummyChoice1)
+        byId <- ledger.flatTransactionById(tree.transactionId, party)
+      } yield {
+        assertEquals("The transaction fetched by identifier does not match", tree, byId)
+      }
+    }
+
+  private[this] val invisibleFlatTransactionById =
+    LedgerTest(
+      "TXInvisibleFlatTransactionById",
+      "Do not expose an invisible flat transaction by identifier") { context =>
+      for {
+        Vector(alpha, beta) <- context.participants(2)
+        party <- alpha.allocateParty()
+        intruder <- beta.allocateParty()
+        dummy <- alpha.create(party, Dummy(party))
+        tree <- alpha.exercise(party, dummy.exerciseDummyChoice1)
+        _ <- synchronize(alpha, beta)
+        failure <- beta.flatTransactionById(tree.transactionId, intruder).failed
+      } yield {
+        assertGrpcError(failure, Status.Code.NOT_FOUND, "Transaction not found, or not visible.")
+      }
+    }
+
+  private[this] val flatTransactionByIdNotFound =
+    LedgerTest(
+      "TXFlatTransactionByIdNotFound",
+      "Return NOT_FOUND when looking up an inexistent flat transaction by identifier") { context =>
+      for {
+        ledger <- context.participant()
+        party <- ledger.allocateParty()
+        failure <- ledger.flatTransactionById("a" * 60, party).failed
+      } yield {
+        assertGrpcError(failure, Status.Code.NOT_FOUND, "Transaction not found, or not visible.")
+      }
+    }
+
+  private[this] val flatTransactionByIdNoParty =
+    LedgerTest(
+      "TXFlatTransactionByIdNotFound",
+      "Return INVALID_ARGUMENT when looking up a flat transaction by identifier without specifying a party") {
+      context =>
+        for {
+          ledger <- context.participant()
+          failure <- ledger.flatTransactionById("not-relevant").failed
+        } yield {
+          assertGrpcError(
+            failure,
+            Status.Code.INVALID_ARGUMENT,
+            "Missing field: requesting_parties")
+        }
+    }
+
+  private[this] val flatTransactionByIdSameAsTransactionStream =
+    LedgerTest(
+      "TXFlatTransactionByIdSameAsTransactionStream",
+      "Expose the same events for each transaction as the output of getTransactions") { context =>
+      for {
+        Vector(alpha, beta) <- context.participants(2)
+        submitter <- alpha.allocateParty()
+        listener <- beta.allocateParty()
+        _ <- alpha.create(submitter, AgreementFactory(listener, submitter))
+        _ <- synchronize(alpha, beta)
+        trees <- alpha.flatTransactions(listener, submitter)
+        _ <- eventually {
```

Comment:
Similar to the comment further above, I don't think we need the `eventually` block here, do we?

Reply:
You are right, probably a leftover from an earlier approach, thanks.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2961/commits/b81eac6b9267070c0f30cb2f09e7fc8444857300